modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -872,262 +847,356 @@ public int getSize() {
+      for (Bytes row : cd.acceptedRows) {
+        m = new Flutation(env, row);
+        for (Entry<Column, Bytes> entry : updates.get(row).entrySet()) {
+          if (isReadLock(entry.getValue())) {
+            m.put(entry.getKey(),
+                ColumnConstants.RLOCK_PREFIX | ReadLockUtil.encodeTs(startTs, true),
+                DelReadLockValue.encodeRollback());
+          } else {
+            m.put(entry.getKey(), ColumnConstants.DEL_LOCK_PREFIX | startTs,
+                DelLockValue.encodeRollback(false, true));
+          }
         }
+        mutations.add(m);
       }
-      mutations.add(m);
+
+      return mutations;
     }
+  }
+
+  class RollbackPrimaryLock extends BatchWriterStep {
 
-    CompletableFuture<Void> cfuture =
-        env.getSharedResources().getBatchWriter().writeMutationsAsyncFuture(mutations);
-    addCallback(cfuture, cd, result -> rollbackPrimaryLock(cd));
+    @Override
+    public Collection<Mutation> createMutations(CommitData cd) {
+      // mark transaction as complete for garbage collection purposes
+      Flutation m = new Flutation(env, cd.prow);
+
+      m.put(cd.pcol, ColumnConstants.DEL_LOCK_PREFIX | startTs,
+          DelLockValue.encodeRollback(startTs, true, true));
+      m.put(cd.pcol, ColumnConstants.TX_DONE_PREFIX | startTs, EMPTY);
+
+      return Collections.singletonList(m);
+    }
   }
 
-  private void rollbackPrimaryLock(CommitData cd) throws Exception {
+  @VisibleForTesting
+  public boolean commitPrimaryColumn(CommitData cd, Stamp commitStamp) {
 
-    // mark transaction as complete for garbage collection purposes
-    Flutation m = new Flutation(env, cd.prow);
+    SyncCommitObserver sco = new SyncCommitObserver();
+    cd.commitObserver = sco;
+    try {
+      CommitStep firstStep = new GetCommitStampStepTest(commitStamp);
 
-    m.put(cd.pcol, ColumnConstants.DEL_LOCK_PREFIX | startTs,
-        DelLockValue.encodeRollback(startTs, true, true));
-    m.put(cd.pcol, ColumnConstants.TX_DONE_PREFIX | startTs, EMPTY);
+      firstStep.andThen(new WriteNotificationsStep()).andThen(new CommitPrimaryStep());
 
-    CompletableFuture<Void> cfuture =
-        env.getSharedResources().getBatchWriter().writeMutationsAsyncFuture(m);
-    addCallback(cfuture, cd,
-        result -> cd.commitObserver.commitFailed(cd.getShortCollisionMessage()));
+      firstStep.compose(cd).thenRun(() -> cd.commitObserver.committed())
+          .exceptionally(throwable -> {
+            cd.commitObserver.failed(throwable);
+            return null;
+          });
+      sco.waitForCommit();
+    } catch (CommitException e) {
+      return false;
+    } catch (Exception e) {
+      throw new FluoException(e);
+    }
+    return true;
   }
 
-  private void beginSecondCommitPhase(CommitData cd, Stamp commitStamp) throws Exception {
-    if (startTs < commitStamp.getGcTimestamp()) {
-      rollbackOtherLocks(cd);
-    } else {
+  class GetCommitStampStep extends CommitStep {
```

Comment:
In order to avoid duplicate code in `GetCommitStampStepTest`, could refactor this class in the following way.

```java
class GetCommitStampStep extends CommitStep {
   protected CompletableFuture<Stamp> getStampOp() {
     return env.getSharedResources().getOracleClient().getStampAsync();
   }
  
   CompletableFuture<Boolean> getMainOp(CommitData cd) {
       return getStampOp().thenApply(commitStamp -> {
￼       writeNotificationsAsync(cd, commitStamp.getTxTimestamp());		￼ +        if (startTs < commitStamp.getGcTimestamp()) {
￼           return false;
￼         } else {
￼           getStats().setCommitTs(commitStamp.getTxTimestamp());
￼           return true;
￼         }
￼       });
   }

}
```
Then in `GetCommitStampStepTest` only need to override `getStampOp()`

Reply:
Did you mean
```
  class GetCommitStampStep extends CommitStep {
    protected CompletableFuture<Stamp> getStampOp() {
      return env.getSharedResources().getOracleClient().getStampAsync();
    }
    
    @Override
    CompletableFuture<Boolean> getMainOp(CommitData cd) {

      return getStampOp().thenApply(commitStamp -> {
        if (startTs < commitStamp.getGcTimestamp()) {
          return false;
        } else {
          getStats().setCommitTs(commitStamp.getTxTimestamp());
          return true;
        }
      });
    }

    @Override
    CompletableFuture<Void> getFailureOp(CommitData cd) {
      return rollbackLocks(cd);
    }

  }
```

`writeNotificationsAsync()` doesn't exist anymore. Also I'm a little confused as to what the difference is between `getGcTimestamp()` and `getTxTimestap()`. The old code used `getGcTimestamp()` to check against `startTs` and then used `getTxTimestamp()` to set the time stamp. I tried to keep it consistent here.

Reply:
I don't know where `writeNotificationsAsync()` came from.  I was just trying to illustrate `getStampOp()` and copied some code from somewhere.  When I was looking at this I had the new and old code open, maybe I copied something from old code.

The Gc timestamp is the Fluo garbage collection timestamp. A transaction can not commit anything older than this time.  The Tx timestamp is the start or commit time stamp.

Reply:
@jkosh44 yeah your example looks correct